readme.md
# Sensors

## Description

The purpose of the project it to demonstrate the implementation of a sensor API.

It is deployed to the following url:

    TODO

## API Routes

### POST /temp

#### Request

The `POST` request at `/temp` should accepts a JSON blob in the following format:

- `{"data": __data_string__}`
- where `__data_string__` is format:
    - `__device_id__:__epoch_ms__:'Temperature':__temperature__`
      - where `__device_id__` is the device ID (int32)
      - `__epoch_ms__` is the timestamp in EpochMS (int64)
      - `__temperature__` is the temperature (float64)
      - and `'Temperature'` is the exact string

- Example `{"data": "365951380:1640995229697:'Temperature':58.48256793121914"}`

#### Response

- If for any reason the data string is not formatted correctly, return `{'error': 'bad request'}` with a `400` status code
- If the temperature is at or over 90
  - return `{'overtemp': True, 'device_id': __device_id__, 'formatted_time': __formatted_time__}`,
    - where `__device_id__` is the device ID (int32)
    - and `__formatted_time__` is the timestamp formatted to `%Y/%m/%d %H:%M:%S`
  - otherwise return `{'overtemp': False}`

### GET /errors

The `GET` request at `/errors` should return all data strings which have not been in the correct format.

### DELETE /errors

The `DELETE` request at `/errors` should clear the errors buffer.

### Live API Documentation

API Documentation can be viewed at the following url:

    TODO

## Releases/Roadmap

See `releases.md` for project progress.

## Development

The local API is deployed to the following url, per `docker-compose.yml`:

http://localhost:3001/

## Deployment

Deploy to Digital Ocean kubernetes cluster via:

    TODO
    
Deployment settings can be found in `./k8/`

## Testing

### Test Suite

    cargo test
    
### Test Coverage

    TODO
    
### Command Line Testing

Select endpoint:

    export URL="http://localhost:3001"
    export URL="http://sensors.kevin.colyar.net"

Errors:

    curl $URL/errors
    curl -X "DELETE" $URL/errors

Temperature:

    curl -X POST -H "Content-Type: application/json" \
    -d '{"data": "365951380:1640995229697:'Temperature':58.48256793121914"}' \
    $URL/temp

    curl -X POST -H "Content-Type: application/json" \
    -d '{"data": "365951380:1640995229697:'Temperature':90.0"}' \
    $URL/temp

    curl -X POST -H "Content-Type: application/json" \
    -d '{"data": ":1640995229697:'Temperature':90.0"}' \
    $URL/temp
    
### Smoke Test

A simple smoke test script of the production API is provided via:

    ./bin/smoke.sh http://localhost:3000

## Questions

Q: Why `int32` for `device_id`, not `uint32`? 

A: Using `int32` per spec, pending future changes.

Q: What unit is temperature in?

A: Assuming fahrenheit based on example data range (e.g. 58.5 & over 90 threshold).

Q: What timezone should `formatted_time` be returned in?

A: Assuming UTC since offset not specified in given format.


releases.md
# Releases

## v0.1.0 - Setup
- [X] Git
- [X] Rust
- [ ] Docker
- [ ] Documentation
- [X] Testing
- [ ] Logging

## v0.2.0 - Routes
- [X] Install Actix
    cargo add actix_web
- [X] Routes - POST /temp
- [X] Routes - GET /errors
- [X] Routes - DELETE /errors

## v0.3.x - Deployment
- [ ] Deployment - k8
- [ ] Deployment - digitalocean
- [ ] Deployment - update pods after image push
  + https://stackoverflow.com/questions/33112789/how-do-i-force-kubernetes-to-re-pull-an-image
  + kubectl rollout restart deployments sensors-deploy
- [ ] Deployment - dns - sensors.kevin.colyar.net

## v0.4.x
- [X] POST temp
- [X] Parse data
- [ ] Response - {'overtemp': False}
- [ ] Response - Error handling - Overtemp
- [ ] Response - Error handling
- [ ] Response - Format timestamp
- [ ] Validation - device id - int32
- [ ] Validation - epoch_ms - int64
- [ ] Validation - temperature - float64

## v0.5.x
- [X] Testing - Setup
- [X] Testing - Unit tests
- [X] Testing - REST client
- [X] Testing - Bad type conversions
- [ ] Testing - payload larger than VARCHAR 255

## v0.6.x
- [ ] Persistence - schema
- [ ] Persistence - stored procedures
- [ ] Persistence - measurements
    + created_at
    + measurement/type
    + device_id 
    + value 
- [ ] Persistence - errors
    + timestamp
    + route
    + method
    + payload
    + sort by timestamp

## v0.7.x
- [ ] Documentation
  + https://www.linode.com/docs/guides/documenting-a-fastapi-app-with-openapi/
- [ ] Documentation - API
  + https://fastapi.tiangolo.com/deployment/docker/#interactive-api-docs

## v0.8.x
- [ ] API versioning

## v1.0.0
- [ ] Push to github
- [ ] Make github project public
- [ ] Archive project
- [ ] Email project

## v1.1.x
- [ ] Linting
- [ ] Coverage

## v1.2.x
- [ ] Persistence - test/staging db
- [ ] Persistence - indices
- [ ] Persistence - measurements - upsert
- [ ] Persistence - errors - save stack trace
- [ ] Persistence - handle db re/connection issues

## v1.3.x
- [ ] Lets Encrypt
- [ ] Kafka
- [ ] Authentication - JWT


tests/endpoints.rs
#[cfg(test)]
mod endpoints {
    use actix_web::{http::header::ContentType, test, App};
    use sensors::endpoints::*;
    use sensors::requests::MeasurementRequest;
    use sensors::responses::{MeasurementValid, MeasurementOvertemp};

    #[actix_web::test]
    async fn test_foo_ok(){
        let app = test::init_service(
            App::new().service(foo_get)
        ).await;

        let req = test::TestRequest::default()
            .uri("/foo")
            .insert_header(ContentType::plaintext())
            .to_request();

        let resp = test::call_service(&app, req).await;

        assert!(resp.status().is_success());
    }

    #[actix_web::test]
    async fn test_temp_ok(){
        let app = test::init_service(
            App::new().service(temp_post)
        ).await;

        let req = test::TestRequest::post()
            .uri("/temp")
            .set_json(
                &MeasurementRequest {
                    data: String::from("365951380:1640995229697:'Temperature':58.48256793121914")
                }
            )
            .to_request();

        let resp = test::call_service(&app, req).await;
        assert!(resp.status().is_success());

        // println!("{:?}", resp);
        // let body = test::read_body(resp).await;
        // println!("{:?}", s);
        // assert_eq!(body, actix_web::web::Bytes::from_static(
        //     b"{\"device_id\":365951380,\"epoch_ms\":1640995229697,\"formatted_time\":\"2022/01/01 00:00:29\",\"name\":\"Temperature\",\"value\":58.48256793121914}"
        // ));
        // assert_eq!(
        //     json!(body),
        //     json!({
        //         "overtemp": false
        //     }))

        let _body: MeasurementValid = test::read_body_json(resp).await;
        // assert_eq!(body, json!(MeasurementOvertemp{
        //     overtemp: true,
        //     device_id: 365951380,
        //     formatted_time: String::from("2022/01/01 00:00:29")
        // }))
    }

    #[actix_web::test]
    async fn test_temp_overtemp(){
        let app = test::init_service(
            App::new().service(temp_post)
        ).await;

        let req = test::TestRequest::post()
            .uri("/temp")
            .set_json(
                &MeasurementRequest {
                    data: String::from("365951380:1640995229697:'Temperature':90.1")
                }
            )
            .to_request();

        let resp = test::call_service(&app, req).await;
        assert!(resp.status().is_success());

        let _body: MeasurementOvertemp = test::read_body_json(resp).await;
    }
}


src/responses.rs
use serde::{Serialize, Deserialize};
use serde_json::{json, Value};

use crate::measurement::Measurement;

#[derive(Serialize, Deserialize)]
pub struct MeasurementValid {
    overtemp: bool
}

#[derive(Serialize, Deserialize)]
pub struct MeasurementOvertemp {
    overtemp: bool,
    device_id: i32,
    formatted_time: String
}

pub fn temperature(m: Measurement) -> Value {
    // Returns json response based on status of temperature
    if m.value >= 90.0 {
        return json!(MeasurementOvertemp {
            overtemp: true,
            device_id: m.device_id,
            formatted_time: m.formatted_time
        })
    }

    return json!(MeasurementValid {
        overtemp: false
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_temperature_valid() {
        let m = Measurement::try_from(
            "365951380:1640995229697:'Temperature':58.48256793121914"
        ).unwrap();
        let resp = temperature(m);

        assert_eq!(resp, json!(MeasurementValid {
            overtemp: false
        }))
    }

    #[test]
    fn test_temperature_over_value() {
        let m = Measurement::try_from(
            "365951380:1640995229697:'Temperature':91.0"
        ).unwrap();
        let resp = temperature(m);

        assert_eq!(resp, json!(MeasurementOvertemp{
            overtemp: true,
            device_id: 365951380,
            formatted_time: String::from("2022/01/01 00:00:29")
        }))
    }
}


src/measurement.rs
use anyhow::Result;
use chrono::prelude::*;
use serde::Serialize;
// use std::str::FromStr;
use std::convert::TryFrom;

#[derive(Debug, Serialize)]
pub struct Measurement {
    pub device_id: i32,
    epoch_ms: i64,
    pub formatted_time: String,
    name: String,
    pub value: f64,
}

// impl FromStr for Measurement {
impl TryFrom<&str> for Measurement {
    // type Err = anyhow::Error;
    type Error = anyhow::Error;

    // fn from_str(s: &str) -> Result<Self> {
    fn try_from(s: &str) -> Result<Self> {
        let split: Vec<_> = s.split(":").collect();

        let device_id: i32 = split[0].parse::<i32>()?;
        let epoch_ms: i64 = split[1].parse::<i64>()?;
        let name: String = split[2].replace("'", "").to_string();
        let value: f64 = split[3].parse::<f64>()?;

        let datetime: DateTime<Utc> =
            DateTime::from_utc(NaiveDateTime::from_timestamp(epoch_ms / 1000, 0), Utc);

        let formatted_time = datetime.format("%Y/%m/%d %H:%M:%S").to_string();

        Ok(Self {
            device_id,
            epoch_ms,
            formatted_time,
            name,
            value,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create() {
        let m =
            Measurement::try_from("365951380:1640995229697:'Temperature':58.48256793121914").unwrap();

        assert!(m.device_id == 365951380);
        assert!(m.epoch_ms == 1640995229697);
        assert!(m.name == "Temperature");
        assert!(m.value == 58.48256793121914);
    }

    #[test]
    fn test_parse_bad_device_id_1() {
        let m = Measurement::try_from(":1640995229697:'Temperature':58.48256793121914");
        assert!(m.unwrap_err().is::<std::num::ParseIntError>());
    }

    #[test]
    fn test_parse_bad_device_id_2() {
        let m = Measurement::try_from("foo:1640995229697:'Temperature':58.48256793121914");
        assert!(m.unwrap_err().is::<std::num::ParseIntError>());
    }

    #[test]
    fn test_negative_epoch_ms() {
        let m =
            Measurement::try_from("365951380:-1640995229697:'Temperature':58.48256793121914").unwrap();

        assert!(m.device_id == 365951380);
        assert!(m.epoch_ms == -1640995229697);
        assert!(m.name == "Temperature");
        assert!(m.value == 58.48256793121914);
    }

    #[test]
    fn test_parse_epoch_ms_1() {
        let m = Measurement::try_from("365951380::'Temperature':58.48256793121914");
        assert!(m.unwrap_err().is::<std::num::ParseIntError>());
    }

    #[test]
    fn test_parse_epoch_ms_2() {
        let m = Measurement::try_from("365951380:foo:'Temperature':58.48256793121914");
        assert!(m.unwrap_err().is::<std::num::ParseIntError>());
    }

    #[test]
    fn test_parse_value_1() {
        let m = Measurement::try_from("365951380:1640995229697:'Temperature':");
        assert!(m.unwrap_err().is::<std::num::ParseFloatError>());
    }

    #[test]
    fn test_parse_value_2() {
        let m = Measurement::try_from("365951380:1640995229697:'Temperature':foo");
        assert!(m.unwrap_err().is::<std::num::ParseFloatError>());
    }

    // #[test]
    // fn test_parse_value_3() {
    //     let m = parse("365951380:1640995229697:'Temperature':-1e1000");
    //     let e = m.unwrap();
    //     println!("value: {}", e.value);
    //     // assert!(e.is::<std::num::ParseFloatError>());
    // }

    // #[test]
    // fn test_parse_value_4() {
    //     let m = parse("365951380:1640995229697:'Temperature':1e100");
    //     assert!(m.unwrap_err().is::<std::num::ParseFloatError>());
    // }

    #[test]
    fn test_formatted_time_1() {
        let m =
            Measurement::try_from("365951380:-1653428445377:'Temperature':58.48256793121914").unwrap();

        assert!(m.formatted_time == "1917/08/10 02:19:15");
    }

    #[test]
    fn test_formatted_time_2() {
        let m =
            Measurement::try_from("365951380:1653428445377:'Temperature':58.48256793121914").unwrap();

        assert!(m.formatted_time == "2022/05/24 21:40:45");
    }
}


src/endpoints.rs
use actix_web::{delete, get, post, web, Responder, HttpResponse, Error};

use crate::measurement::*;
use crate::requests::{MeasurementRequest};
use crate::responses::{temperature};

#[post("/temp")]
pub async fn temp_post(item: web::Json<MeasurementRequest>) -> impl Responder {
    // let m = Measurement::try_from("365951380:1640995229697:'Temperature':58.48256793121914");
    // println!("item: {:?}", &item);
    let m = Measurement::try_from(&*item.data);

    return web::Json(temperature(m.unwrap()))
}

#[get("/foo")]
pub async fn foo_get() -> Result<HttpResponse, Error> {
    let m = Measurement::try_from("365951380:1640995229697:'Temperature':58.48256793121914");

    // let res = HttpResponse::Ok().finish

    // web::Json(m.unwrap())
     Ok(HttpResponse::Ok().json(web::Json(m.unwrap())))
}

#[get("/errors")]
pub async fn errors_get() -> impl Responder {
    format!("GET /errors")
}

#[delete("/errors")]
pub async fn errors_delete() -> impl Responder {
    format!("GET /errors")
}
